### PR TITLE
adding shared_eden_dataloader.

### PR DESCRIPTION
**Pull request by yzhang123**
### Description

SharedEdenDataloader 
Internal PR of https://github.com/NVIDIA/bionemo-framework/pull/1091
Contributed by Basecamp Research  @Geraldene

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
# TODO: Add code snippet
```

### Pre-submit Checklist

- [x] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [x] I have added/updated tests as needed
- [ ] All existing tests pass successfully

## Summary by CodeRabbit

- New Features
  - High-performance sharded data loader with precomputed window DBs, per-sample sequence stores, CLI precompute, RC augmentation, window-access logging, and PyTorch Lightning integration.
  - Expanded model support: Eden-flavored Llama configs with extended-context options and new LLAMA model sizes; training CLI and run naming updated to support LLAMA.

- Documentation
  - Added user guide describing sharded data loader layout, preprocessing, and usage.

- Tests
  - Added comprehensive dataset/data‑module tests and end-to-end Llama pretrain/finetune tests.

- Chores
  - Updated license header year.

### CODE REVIEW

File: sub-packages/bionemo-evo2/src/bionemo/evo2/data/sharded_eden_dataloader.py [hunk 1]
```diff
@@ -0,0 +1,937 @@
...
+_, HAVE_TE = safe_import("transformer_engine")
+
+if TYPE_CHECKING:
+    from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
+
+
+def extract_sample_id(sequence_id: str) -> str:
+    """Extract sample ID from sequence ID format: BCR__EXT-SAMPLE1__CT1-1."""
+    parts = sequence_id.split("__")[1].split("-")[1:]
+    return ".".join(parts)
```
[the remark this answers is not in the document]
Comment (tshimko-nv) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/data/sharded_eden_dataloader.py:64]: This is probably worth documenting somewhere or raising a more helpful error to guide the user in naming convention.

**Review by tshimko-nv — APPROVED** [review]
One comment on clarifying naming conventions, otherwise looks good.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/data/sharded_eden_dataloader.py [hunk 2]
```diff
@@ -0,0 +1,937 @@
...
+        global_batch_size: int = 4,
+        rampup_batch_size: Optional[List[int]] = None,
+        num_workers: int = 8,
+        pin_memory: bool = True,
+        persistent_workers: bool = False,
+        create_attention_mask: bool = False,
+        vocab_file: Optional[str] = None,
+        merges_file: Optional[str] = None,
+        rc_aug: bool = False,
+        stride: int = 7992,
```
Comment (balvisio) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/data/sharded_eden_dataloader.py:94]: The `seq_length` and `stride` are saved in the metadata table of the window database. For consistency, I think these two parameters should be read from the input db files instead of being provided by the user as input params.

Comment (balvisio) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/data/sharded_eden_dataloader.py:94]: Ah, there is already a check in `ShardedEdenDataloader` that accounts for this:
```
 if db_window_size != self.seq_length or db_stride != self.stride:
                raise ValueError(
                    f"Database metadata mismatch! "
                    f"DB created with window_size={db_window_size}, stride={db_stride}. "
                    f"Dataset configured with seq_length={self.seq_length}, stride={self.stride}. "
                    f"Please re-run pre-computation or check your config."
                )
```

But to me doesn't make too much sense to include a parameter that, in reality, can only take only one valid value. :)

File: sub-packages/bionemo-evo2/src/bionemo/evo2/models/llama.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    # wavelen < high_freq_wavelen: do nothing
+    # wavelen > low_freq_wavelen: divide by factor
+    inv_freq_llama = torch.where(wavelen > low_freq_wavelen, inv_freq / factor, inv_freq)
+    # otherwise: interpolate between the two, using a smooth factor
+    smooth_factor = (old_context_len / wavelen - low_freq_factor) / (high_freq_factor - low_freq_factor)
+    smoothed_inv_freq = (1 - smooth_factor) * inv_freq_llama / factor + smooth_factor * inv_freq_llama
+    is_medium_freq = ~(wavelen < high_freq_wavelen) * ~(wavelen > low_freq_wavelen)
+    inv_freq_llama = torch.where(is_medium_freq, smoothed_inv_freq, inv_freq_llama)
+
+    return inv_freq_llama
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/models/llama.py]: We can import this from nemo: https://github.com/NVIDIA-NeMo/NeMo/blob/main/nemo/collections/llm/gpt/model/llama.py#L1330

**Review by jstjohn — APPROVED** [review]

**Comment by jstjohn** [comment]
Note, this PR addresses feedback left on #1091

File: sub-packages/bionemo-evo2/src/bionemo/evo2/models/llama.py (hunk 3 above)
Comment (yzhang123) [reply sub-packages/bionemo-evo2/src/bionemo/evo2/models/llama.py]: replaced
